lib/rex/proto/ldap/server.rb

```diff
@@ -149,104 +153,175 @@ def dispatch_request(cli, data)
+                      result_code = Net::LDAP::ResultCodeNoSuchObject
+                    else
+                      client.write(res)
+                      result_code = Net::LDAP::ResultCodeSuccess
+                    end
+                    processed_pdu_data[:pdu_type] = pdu.app_tag
+                    encode_ldap_response(
+                      pdu.message_id,
+                      result_code,
+                      '',
+                      Net::LDAP::ResultStrings[result_code],
+                      Net::LDAP::PDU::SearchResult
+                    )
+                  when Net::LDAP::PDU::UnbindRequest
+                    client.close
+                    nil
+                  else
+                    if suitable_response(pdu.app_tag)
+                      result_code = Net::LDAP::ResultCodeUnwillingToPerform
+                      encode_ldap_response(
+                        pdu.message_id,
+                        result_code,
+                        '',
+                        Net::LDAP::ResultStrings[result_code],
+                        suitable_response(pdu.app_tag)
+                      )
+                    else
+                      client.close
+                    end
+                  end
+
+            if @pdu_process[pdu.app_tag] && !processed_pdu_data.empty?
+              @pdu_process[pdu.app_tag].call(processed_pdu_data)
+            end
+            send_response(client, res) unless res.nil?
           rescue StandardError => e
             elog(e)
-            cli.close
+            client.close
             raise e
           end
         end
 
         #
         # Encode response for LDAP client consumption
         #
-        # @param msgid [Integer] LDAP message identifier
-        # @param code  [Integer] LDAP message code
-        # @param dn    [String]  LDAP distinguished name
-        # @param msg   [String]  LDAP response message
-        # @param tag   [Integer] LDAP response tag
+        # @param msgid        [Integer] LDAP message identifier
+        # @param code         [Integer] LDAP message code
+        # @param dn           [String]  LDAP distinguished name
+        # @param msg          [String]  LDAP response message
+        # @param tag          [Integer] LDAP response tag
+        # @param context_data [String]  Additional data to serialize in the sequence
+        # @param context_code [Integer] Context Specific code related to `context_data`
         #
         # @return [Net::BER::BerIdentifiedOid] LDAP query response
-        def encode_ldap_response(msgid, code, dn, msg, tag)
+        def encode_ldap_response(msgid, code, dn, msg, tag, context_data = nil, context_code = nil)
+          tag_sequence = [
+            code.to_ber_enumerated,
+            dn.to_ber,
+            msg.to_ber
+          ]
+
+          if context_data && context_code
+            tag_sequence << [context_data.to_ber].to_ber_contextspecific(context_code)
```

Comment:
Extra context data should not be wrapped in an Array object.  With this additional adjustment NTLM auth can succeed.
```suggestion
            tag_sequence << context_data.to_ber_contextspecific(context_code)
```

From client side after change:
```
$ ldapsearch -H ldap://server.example.com -Y ntlm -U admin -b 'dc=server,dc=example,dc=com'
SASL/NTLM authentication started
Please enter your password:
ldap_sasl_interactive_bind: Can't contact LDAP server (-1)
        additional info: saslBindInProgress
```

With change and all others from previous comments to support `NTLM` as mechanism referenced:
```
msf6 auxiliary(server/capture/ldap) > run

[*] Server started.
[+] LDAP Login Attempt => From:127.0.0.1:58642   Username:admin  ntlm_hash::e1636396de9664cd2ff248fa8f561615747fb96d869af536     Domain:EXAMPLE
```